Source/i_midipipe.c

```diff
@@ -0,0 +1,513 @@
+    ok = WritePipe(packet);
+    NET_FreePacket(packet);
+
+    ok = ok && BlockForAck();
+    if (!ok)
+    {
+        DEBUGOUT("I_MidiPipe_StopSong failed");
+        return;
+    }
+
+    DEBUGOUT("I_MidiPipe_StopSong succeeded");
+}
+
+//
+// I_MidiPipe_ShutdownServer
+//
+// Tells the MIDI subprocess to shutdown.
+//
+void I_MidiPipe_ShutdownServer()
+{
+    boolean ok;
+    net_packet_t *packet;
+
+    packet = NET_NewPacket(2);
+    NET_WriteInt16(packet, MIDIPIPE_PACKET_TYPE_SHUTDOWN);
+    ok = WritePipe(packet);
+    NET_FreePacket(packet);
+
+    ok = ok && BlockForAck();
+    FreePipes();
+
+    midi_server_initialized = false;
+
+    if (!ok)
+    {
+        DEBUGOUT("I_MidiPipe_ShutdownServer failed");
+        return;
+    }
+
+    DEBUGOUT("I_MidiPipe_ShutdownServer succeeded");
+}
+
+//=============================================================================
+//
+// Public Interface
+//
+
+//
+// I_MidiPipeInitServer
+//
+// Start up the MIDI server.
+//
+boolean I_MidiPipe_InitServer()
+{
+    TCHAR dirname[MAX_PATH + 1];
+    DWORD dirname_len;
+    char *module = NULL;
+    char *cmdline = NULL;
+    char *fullpath = NULL;
+    char params_buf[128];
+    SECURITY_ATTRIBUTES sec_attrs;
+    PROCESS_INFORMATION proc_info;
+    STARTUPINFO startup_info;
+    BOOL ok;
+
+    if (!UsingNativeMidi() /*|| strlen(snd_musiccmd) > 0*/)
+    {
+        // If we're not using native MIDI, or if we're playing music through
+        // an exteranl program, we don't need to start the server.
+        return false;
+    }
+
+    // Get directory name
+    memset(dirname, 0, sizeof(dirname));
+    dirname_len = GetModuleFileName(NULL, dirname, MAX_PATH);
+    if (dirname_len == 0)
+    {
+        return false;
+    }
+    RemoveFileSpec(dirname, dirname_len);
+
+    // Define the module.
+    module = "woof-midiproc.exe";
+
+    // Set up pipes
+    memset(&sec_attrs, 0, sizeof(SECURITY_ATTRIBUTES));
+    sec_attrs.nLength = sizeof(SECURITY_ATTRIBUTES);
+    sec_attrs.bInheritHandle = TRUE;
+    sec_attrs.lpSecurityDescriptor = NULL;
+
+    if (!CreatePipe(&midi_process_in_reader, &midi_process_in_writer, &sec_attrs, 0))
+    {
+        DEBUGOUT("Could not initialize midiproc stdin");
+        return false;
+    }
+
+    if (!SetHandleInformation(midi_process_in_writer, HANDLE_FLAG_INHERIT, 0))
+    {
+        DEBUGOUT("Could not disinherit midiproc stdin");
+        return false;
+    }
+
+    if (!CreatePipe(&midi_process_out_reader, &midi_process_out_writer, &sec_attrs, 0))
+    {
+        DEBUGOUT("Could not initialize midiproc stdout/stderr");
+        return false;
+    }
+
+    if (!SetHandleInformation(midi_process_out_reader, HANDLE_FLAG_INHERIT, 0))
+    {
+        DEBUGOUT("Could not disinherit midiproc stdin");
+        return false;
+    }
+
+    // Define the command line.  Version, Sample Rate, and handles follow
+    // the executable name.
+    int snd_samplerate = 44100;
```

Comment:
No variable declaration in code, please.

Reply:
And why do we have to hard-code this here?

Reply:
snd_samplerate is a config parameter in Crispy Doom, should I replace it here with a constant?

Reply:
Ah, right, it's even hard-coded to that value in Woof's own `i_sound.c`.

Reply:
Anyway, please move the variable declaration so it doesn't appear in the middle of the code.